packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts

```diff
@@ -315,14 +291,105 @@ function checkFunctionExpressionReturnType(
     node.type === AST_NODE_TYPES.ArrowFunctionExpression &&
     returnsConstAssertionDirectly(node)
   ) {
+    return true;
+  }
+
+  return false;
+}
+
+/**
+ * Check that the function expression or declaration is valid.
+ */
+function isValidFunctionReturnType(
+  node:
+    | TSESTree.ArrowFunctionExpression
+    | TSESTree.FunctionDeclaration
+    | TSESTree.FunctionExpression,
+  options: Options,
+  isParentCheck = false,
+): boolean {
+  if (
+    !isParentCheck &&
+    options.allowHigherOrderFunctions &&
+    doesImmediatelyReturnFunctionExpression(node)
+  ) {
+    return true;
+  }
+
+  if (node.returnType || isConstructor(node.parent) || isSetter(node.parent)) {
+    return true;
+  }
+
+  return false;
+}
+
+/**
+ * Checks if a function declaration/expression has a return type.
+ */
+function checkFunctionReturnType(
+  node:
+    | TSESTree.ArrowFunctionExpression
+    | TSESTree.FunctionDeclaration
+    | TSESTree.FunctionExpression,
+  options: Options,
+  sourceCode: TSESLint.SourceCode,
+  report: (loc: TSESTree.SourceLocation) => void,
+): void {
+  if (isValidFunctionReturnType(node, options)) {
+    return;
+  }
+
+  report(getReporLoc(node, sourceCode));
+}
+
+/**
+ * Checks if a function declaration/expression has a return type.
+ */
+function checkFunctionExpressionReturnType(
+  node: TSESTree.ArrowFunctionExpression | TSESTree.FunctionExpression,
+  options: Options,
+  sourceCode: TSESLint.SourceCode,
+  report: (loc: TSESTree.SourceLocation) => void,
+): void {
+  if (isValidFunctionExpressionReturnType(node, options)) {
     return;
   }
 
   checkFunctionReturnType(node, options, sourceCode, report);
 }
 
+/**
+ * Check whether any ancestor of the provided node has a valid return type, with
+ * the given options.
+ */
+function ancestorHasReturnType(
+  ancestor: TSESTree.Node | undefined,
+  options: Options,
+): boolean {
+  let hasType = false;
+
+  while (ancestor) {
+    switch (ancestor.type) {
+      case AST_NODE_TYPES.ArrowFunctionExpression:
+      case AST_NODE_TYPES.FunctionExpression:
+        hasType =
+          isValidFunctionExpressionReturnType(ancestor, options) ||
+          isValidFunctionReturnType(ancestor, options, true);
+        break;
+      case AST_NODE_TYPES.FunctionDeclaration:
+        hasType = isValidFunctionReturnType(ancestor, options, true);
+        break;
+    }
+
+    ancestor = ancestor.parent;
+  }
+
+  return hasType;
+}
```

Comment:
Almost correct, I can see two problems with this:

1.

You should exit this loop if:
- you see anything that's not a `ReturnStatement`.
- you see a function declaration.

Examples that will currently behave incorrectly:

```ts
function a(): void {
  function b() {} // not returned, shouldn't care about the parent fn's return type
  const x = () => {}; // not returned, shouldn't care about the parent fn's return type
  (function () {}); // not returned, shouldn't care about the parent fn's return type

  function c() {
    return () => {}; // returned from inner function - should not be affected by outer function's return type
  }

  return;
}
```

2.

This is will return whatever the outer-most function's return type is, and ignore any functions it encounters along the way.

Eg, this will return false, even though it encountered a return type in function `b`:
```ts
function a() /* no return */ {
  return function b(): () => void {
    return function c() {}
  }
}
```

You probably want to hard-exit when `hasType === true`

Reply:
> You probably want to hard-exit when `hasType === true`

I added a couple more tests from the docs and breaking errors came from the point you made in 2. Breaking early fixed all of those. 

> You should exit this loop if:
>
> - you see anything that's not a ReturnStatement.
> - you see a function declaration.

I couldn't replicate what you mentioned in number 1. The code passed fine in the tests and exiting when not a return statement broke a lot of other tests.

Since this new method augments the previous check, it was only meant to catch stuff that was previously missed. The new failures I found were also failing beforehand (I think).

I can add even more tests. I was going through the docs and adding test cases to check my assumptions.

Reply:
Okay see my latest comment. I think I figured it out.